src/orion/core/worker/warm_start/multi_task_wrapper.py

```diff
@@ -0,0 +1,249 @@
+"""Wrapper that makes an algorithm warm-starteable by observing trials from different tasks.
+
+The wrapper adds a `task_id` dimension to the search space, and uses the value of `task_id` to
+differentiate between the current "target" experiment (`task_id==0`) and the other experiments
+from the knowledge base (`task_id>0`).
+
+The trials from the knowledge base that can fit within the same space are reused and given a
+different value for a new "task_id" dimension.
+"""
+from __future__ import annotations
+
+import copy
+from collections import defaultdict
+from contextlib import contextmanager
+from logging import getLogger
+
+from orion.algo.space import Categorical, Space
+from orion.core.utils.format_trials import dict_to_trial
+from orion.core.worker.algo_wrappers.algo_wrapper import AlgoT
+from orion.core.worker.algo_wrappers.transform_wrapper import (
+    TransformWrapper,
+    _copy_status_and_results,
+)
+from orion.core.worker.trial import Trial
+from orion.core.worker.warm_start.knowledge_base import ExperimentConfig, KnowledgeBase
+from orion.core.worker.warm_start.warm_starteable import WarmStarteable
+
+logger = getLogger(__file__)
+
+TARGET_TASK_ID = 0
+
+# pylint: disable=too-many-public-methods
+class MultiTaskWrapper(TransformWrapper[AlgoT], WarmStarteable):
+    """Wrapper that makes the algo "multi-task" by adding a task id to the inputs."""
+
+    def __init__(
+        self,
+        space: Space,
+        algorithm: AlgoT,
+    ):
+        super().__init__(space=space, algorithm=algorithm)
+        self.current_task_id: int = TARGET_TASK_ID
+        self._total_warm_start_trials = 0
+        self._max_trials: int | None = None
+
+    def transform(self, trial: Trial) -> Trial:
+        return self._add_task_id(trial, self.current_task_id)
+
+    def reverse_transform(self, trial: Trial) -> Trial:
+        return self._remove_task_id(trial)
+
+    # pylint: disable=arguments-differ
+    @classmethod
+    def transform_space(cls, space: Space, knowledge_base: KnowledgeBase) -> Space:
+        """Transform an (outer) space, returning the (inner) space of the wrapped algorithm.
+
+        This returns what will become `self.algorithm.space` in an instance of this class.
+        """
+        if "task_id" in space:
```

Comment:
I think this should be tested.

Reply:
Fixed in 47d8a951b3953e4472cf5d0a7e3d4c64c1af7b72